Better error message on unrecognized command

sbruce opened the pull request:
License: MIT
Signed-off-by: Shaun Bruce <shaun.m.bruce@gmail.com>

whyrusleeping commented on commands/cli/parse.go:
```diff
@@ -227,11 +229,61 @@ func parseArgs(inputs []string, stdin *os.File, argDefs []cmds.Argument, recursi
 	// if we have more arg values provided than argument definitions,
 	// and the last arg definition is not variadic (or there are no definitions), return an error
 	notVariadic := len(argDefs) == 0 || !argDefs[len(argDefs)-1].Variadic
 	if notVariadic && len(inputs) > len(argDefs) {
```
I wonder if there are any legitimate cases where we would want to print this.. Or if this always means 'you spelled it wrong'.

whyrusleeping commented on commands/cli/parse.go:
```diff
@@ -227,11 +229,61 @@ func parseArgs(inputs []string, stdin *os.File, argDefs []cmds.Argument, recursi
 		numInputs += 1
 	}
 
+	searchUnknownCmd := func(args []string) []string {
```
maybe move this function out into its own file in the same directory, as well as the sorting code below.

sbruce commented:
Thanks for the feedback. I moved the code to a new file. I also renamed the **searchUknownCmd** func to **suggestUnknownCmd**. That name sounded a little better.

jbenet commented on commands/cli/cmd_suggestion.go:
```diff
@@ -0,0 +1,71 @@
+	"strings"
+
+	cmds "github.com/ipfs/go-ipfs/commands"
+	levenshtein "github.com/texttheater/golang-levenshtein/levenshtein"
```
will want to vendor. run 
```
make vendor
```
from root.

---

in the event that godeps barfs at you with all sorts of requests, try `go get` on all the needed packages. 

people report that this works too:
```
godep restore
```
or even setting
```
export GOPATH=`pwd`/Godeps/_workspace:$GOPATH
```

jbenet commented on commands/cli/parse.go:
```diff
@@ -231,7 +231,15 @@ func parseArgs(inputs []string, stdin *os.File, argDefs []cmds.Argument, recursi
+		if len(suggestions) > 1 {
+			return nil, nil, fmt.Errorf("Unknown Command \"%s\"\n\nDid you mean any of these?\n\n\t%s", inputs[0], strings.Join(suggestions, "\n\t"))
+		} else if len(suggestions) > 0 {
+			return nil, nil, fmt.Errorf("Unknown Command \"%s\"\n\nDid you mean this?\n\n\t%s", inputs[0], suggestions[0])
```
maybe just:

```
Did you mean?
<cmd>
```

jbenet commented:
as someone liable to type with one hand, i find this nice. I'm curious what other people think though and whether this is stuff to put in ipfs, or to leave up to shells?

whyrusleeping commented:
my golden rule: "what does git do?"

```
whyrusleeping@idril ~> git stats
git: 'stats' is not a git command. See 'git --help'.

Did you mean this?
	status
```

whyrusleeping commented:
@sbruce if you want, one of us can probably handle the godep vendoring. Its quite the obnoxious process

sbruce commented:
@whyrusleeping Thanks, but I think I finally figured out godeps.

jbenet commented:
@sbruce _almost_ there. need to run `make vendor` (which runs: `godep save -r ./...` and rewrites the import paths to point to our workspace)

sbruce commented:
Okay, make vendor should be fixed.  I also squashed, and rebased with master.

whyrusleeping commented:
LGTM

jbenet commented:
@sbruce i should've brought this up. could you write a test for it? if people start depending on this, i want to make sure it works. a sharness test would do. this should get you started:

```
cat <the-below> >test/sharness/t0150-clisuggest.sh
chmod +x test/sharness/t0150-clisuggest.sh
```
```sh
#!/bin/sh

test_description="Test ipfs cli cmd suggest"

. lib/test-lib.sh

test_suggest() {


	test_expect_success "test command fails" '
		test_must_fail ipfs <cmd> >actual
	'

	test_expect_success "test one command is suggested" '
		grep "<cmd1>" actual ||
		fsh cat actual
	'

	test_expect_success "test command fails" '
		test_must_fail ipfs <cmd> >actual
	'

	test_expect_success "test multiple commands are suggested" '
		grep "<cmd1>" actual &&
		grep "<cmd2>" actual ||
		fsh cat actual
	'

}

test_init_ipfs

test_suggest

test_launch_ipfs_daemon

test_suggest

test_kill_ipfs_daemon

test_done
```

sbruce commented:
A sharness test has been added. Let me know if you think it isn't thorough enough.

jbenet commented:
@sbruce sorry for the delay. I missed the test update. this LGTM. could you rebase on master? i'd be happy to include it in 0.3.6 if it's rebased today (else 0.3.7).

jbenet commented:
and thanks! :+1: :smile:

sbruce commented:
I just rebased to master and pushed the branch. It should hopefully be ready to go.

And, you're welcome. I'm looking forward to helping out more.

jbenet commented:
thanks @sbruce ! :+1:
